CompareRaftMembers/DiffStreamOplogFilter.js

```diff
@@ -0,0 +1,240 @@
+            raftSessionState.oplogKeys = new Set();
+            raftSessionState.inputBuffer = [];
+        }
+    }
+
+    _setupOplogStream(raftSessionState, oplogStream) {
+        // no 'end' event handling is required since the oplog streams never end
+        oplogStream
+            .on('data', data => this._onOplogEvent(raftSessionState, data))
+            .on('error', err => {
+                this.emit('error', err);
+            });
+    }
+
+    _onOplogEvent(raftSessionState, data) {
+        const { entry } = data;
+        if (entry === null) {
+            // a null entry means we reached the end of
+            // the raft oplog, this is the point where
+            // it becomes safe to process the raft
+            // session's current input buffer
+            this._processInputBuffer(raftSessionState);
+        } else if (entry.method === 'BATCH') {
+            // only consider events from BATCH records
+            const fullKey = `${entry.bucket}/${entry.key}`;
+            raftSessionState.oplogKeys.add(fullKey);
+        }
+    }
+
+    _transform(diffEntry, encoding, callback) {
+        const bucketName = this._extractBucketName(diffEntry);
+        async.waterfall([
+            next => {
+                const rsId = this.bucketNameToRaftSessionId[bucketName];
+                if (rsId !== undefined) {
+                    return next(null, rsId);
+                }
+                return this._fetchBucketRaftSessionId(bucketName, next);
+            },
+        ], (err, rsId) => {
+            if (err) {
+                this.emit('error', err);
+                return callback();
+            }
+            // bucket not found: ignore the entry
+            if (rsId === null) {
+                return callback();
+            }
+            const raftSessionState = this.raftSessionStates[rsId];
+            // if for some reason the returned raft session ID is not
+            // managed by us (e.g. a bucket has been recently
+            // recreated with a different raft session ID), ignore the
+            // entry
+            if (!raftSessionState) {
+                return callback();
+            }
+            raftSessionState.inputBuffer.push(diffEntry);
+            this.nBufferedEntries += 1;
+            if (this.nBufferedEntries !== this.maxBufferedEntries) {
+                // we can accept more input
+                return callback();
+            }
+            // wait until the current read buffer is flushed before
+            // accepting more input
+            this.pendingTransformCallback = callback;
+            return undefined;
+        });
+    }
+
+    _flush(callback) {
+        if (this.nBufferedEntries === 0) {
+            // end the output stream if there are no entries to forward
+            this.push(null);
+            return callback();
+        }
+        this.pendingFlushCallback = callback;
+        return undefined;
```

Comment:
why the explicit undefined return?

Reply:
one of our linter rules is `consistent-return`, which is making sure that if we return somewhere (even `undefined`, here at https://github.com/scality/s3utils/pull/246/files#diff-8275bd7edd92cb9b8c115839e21be4e42f8d5e708502eff30fcf85566bd6793dR137), we have to return something everywhere in the function.

Reply:
We abuse the `return callback()` notation though, since we don't care about the returned value.